### ISSUE TITLE
Client list in GUI has no scroll bar (overflow is hidden)

### ISSUE THREAD

**Issue by chris-counteractive** [Velocidex/velociraptor#357]
When you use the top search bar and get a list of clients that overflows the UI view, the list is truncated: that is, the overflow is hidden, with no scroll bar:

![broken](https://user-images.githubusercontent.com/48190665/81596442-b2ee8480-9389-11ea-92a6-5fe47f31312a.jpg)

**Version:** [release 0.4.2](https://github.com/Velocidex/velociraptor/releases/tag/v0.4.2)
**Server OS:** Ubuntu 18.04
**Client OS:** Windows 10
**Browser:** reproduced in latest versions of Chrome and Firefox

We've been working around it by adding `overflow-y: auto;` to the ancestor `div[ui-view]` selector in the browser:

```css
div[ui-view] {
    width: 100%;
    height: 100%;
    overflow-y: auto;
}
```

There may be a better place to fix that, updating `div[ui-view]` in `base.css` (below) might break something somewhere else, but even adding a helper class would be a quick fix.

https://github.com/Velocidex/velociraptor/blob/cee9c79ccb3008f20f0d98a5171415226c494f0b/gui/static/css/base.css#L157-L160

Happy to submit a PR if you prefer.  Thanks for an **amazing** tool.

**Comment by scudette** [comment]
Huh - I think I deliberately disabled the scrolling. Maybe that's how i
did it?
The trouble is that the API returns 50 clients in arbitrary order so
scrolling does not make sense (there is no order so you would just see
the same clients over and over for each screen refresh). The GUI widget
is infinitely scrolled so it just keeps going - every screen refresh
takes the next 50 in the list. So the scrolling will stop for the total
clients but the clients may be duplicated in each screen. Pretty weird
experience ...
The reason is that there is no sorting on clients (you can think of
100k clients - it might take too long to sort that list on e.g.
hostname but might be ok to sort on client id).
I know people have asked for the scroll bar back in the past and it
might be useful. I think a better solution than limiting the size of
the screen refresh is to limit the time - if it takes more than a
couple sec to do a screen refresh dont worry about it?
Typically we try to search better for the clients (you can use
wildcards in the search box).
I also think it might be nice to replace that table with the standard
Velociraptor table (so it is sortable and searchable). The standard
table is not infinite so the size can be limited (by default 500) 
What do you think?
ThanksMike

On Mon, 2020-05-11 at 11:21 -0700, Chris Hendricks wrote:
> When you use the top search bar and get a list of clients that
> overflows the UI view, the list is truncated: that is, the overflow
> is hidden, with no scroll bar:
> 
> 
> 
> Version: release 0.4.2
> 
> Server OS: Ubuntu 18.04
> 
> Client OS: Windows 10
> 
> Browser: reproduced in latest versions of Chrome and Firefox
> 
> We've been working around it by adding overflow-y: auto; to the
> ancestor div[ui-view] selector in the browser:
> 
> div[ui-view] {    width: 100%;    height: 100%;    overflow-y: auto;}
> 
> There may be a better place to fix that, updating div[ui-view] in
> base.css (below) might break something somewhere else, but even
> adding a helper class would be a quick fix.
> 
> https://github.com/Velocidex/velociraptor/blob/cee9c79ccb3008f20f0d98a5171415226c494f0b/gui/static/css/base.css#L157-L160
> 
> Happy to submit a PR if you prefer.  Thanks for an amazing tool.
> 
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> 
> [
> {
> "@context": "http://schema.org",
> "@type": "EmailMessage",
> "potentialAction": {
> "@type": "ViewAction",
> "target": "https://github.com/Velocidex/velociraptor/issues/357",
> "url": "https://github.com/Velocidex/velociraptor/issues/357",
> "name": "View Issue"
> },
> "description": "View this Issue on GitHub",
> "publisher": {
> "@type": "Organization",
> "name": "GitHub",
> "url": "https://github.com"
> }
> }
> ]
-- 
Mike Cohen
Digital Paleontologist

Velocidex Enterprises
https://www.velocidex.com

+61.470238491

**Comment by chris-counteractive** [comment]
Thanks for the reply, sir!  I suspected there was more to the story.  I definitely see the design conundrum, and I respect the tradeoff.  A couple thoughts:

1. In this case (a wildcard search or "all"), it's not obvious the limited list is a design decision, which makes it feel like a bug.  If there's 49 clients returned, so 49 in the list, only maybe 15-30 show on your average screen ... necessitating that `overflow-y: auto` even if you keep the current approach and have <50 matches.  And of course, typing "all" makes it natural to expect _all_ of the clients in the list.
1. If all you have is the 50 arbitrary clients from the API, a label could make that clear: `50 of the 312 matching clients`, or similar, but that just highlights the dissonance between the search term and the results.  This doesn't seem like a great choice.
1. In other cases where the app limits each pull, the typical idiom is pagination, which I personally don't love ... and in any case, would probably require server-side sort and a stable return order, so not a great option.
1. For some of us (e.g., consultants in unfamiliar networks) discovery is a key UI benefit -- we often don't know what hostnames to expect or search for -- and a full list is worth the expense.  Currently we pull the list from the `Server.Information.Clients` server artifact and use it to browse hostnames, or use some command-line kung fu ... not ideal. Could be there's a feature that we just don't know about, in which case I apologize for the user error.
1. If you wanted it optional, you could hide list-construction behind a button (`build full list`) and consider caching on the client and/or server to keep the resource burden lower.
    1. In most environments I expect this would be pretty low-cost, you might forego the button if it's less than some threshold row count (or use your screen refresh timeout idea)
    1. 100,000 rows of json of `Server.Information.Clients` data would be the better part of 20mb, but only ~25% of that gzipped on the wire, not terrible.
    1. I'm not sure if most client-side table sorting/paginating libraries would tip over on data of that size, but I bet we could find one that works well.  I'll look into it.
1. If you still prefer the search-first idiom, I'd recommend the search bar be much more flexible and inclusive.  In situations where you don't know the hostname, it doesn't search the other useful fields you could choose to show (i.e., you can't search on OS or revision or labels or last IP ...).  I think a client-side filter would probably perform fine for this too, but I'll have to test that assumption.

I appreciate the opportunity, and I'll do a bit more experimenting, thanks again for the quick response!

**Comment by scudette** [comment]
Thanks for the feedback. I thought more about it and the current UI design makes it hard to move away from the infinite table (there is a label selector). As you pointed out the viewport may not even have 50 results. I definitely think your CSS tweak is the right thing to do right now.

As for the wider question of filtering by clients - your approach of running the server artifact `Server.Information.Clients` is definitely one way to go - it only takes a couple seconds to run and allows one to post process the list, filter etc.

You do not actually need to do any command line magic if you use the notebook. You can use the notebook to post process the specific server collection of each `Server.Information.Clients` you ran but this is not as convenient because you probably want to run this over and over to see how clients are joining over time.

Since an artifact is just a way to encapsulate VQL you just can grab the original VQL from the artifact definition and just paste the VQL into a VQL cell. Then every time you re-calculate the cell it will update the client collection

![image](https://user-images.githubusercontent.com/3856546/81636420-2a3e1a00-9457-11ea-91e8-48d83ec20570.png)

You can then directly filter/search etc on any field you want.

**Comment by chris-counteractive** [comment]
Thank you - that's great advice regarding the notebooks, it'll see use tomorrow!

To close the loop and satisfy my own curiosity, I put together a small demo to test client-side data filtering at https://github.com/chris-counteractive/table-sort-test. I cranked out [100,000 rows](https://github.com/chris-counteractive/table-sort-test/blob/master/fake-clients-100000.json) of data using [faker](https://faker.readthedocs.io/en/master/) and chose [datatables](https://datatables.net/) as an example of a fully-featured table client-side table library.

TL;DR: it takes ~7 seconds for first render of the table, subsequent sorts are less than a second, you get client-side pagination with instant page-size changes, and the search-box filtering is basically instant across all fields.  Transferring the json doesn't move the needle on the overall time unless you're on a very slow link.

It's not a drop-in replacement, of course: that first long render blocks the thread which is janky, but it shows promise.  These may be things you know or have already considered and dismissed, but it helped me validate some mental models.  Thanks again for the conversation, always happy to help if you're open to contributions.

**Comment by scudette** [comment]
This should be fixed in 0.4.4
